### ISSUE TITLE
NA/M5 addmm / matmul on MPS ~10–20% slower than PyTorch for 1280×1280 BF16 (hurts Nano Chat training)

### ISSUE THREAD

**Issue by Anemll** [ml-explore/mlx#3196]
Summary
On an Mac M5, mlx.addmm and A @ B are consistently ~1.1–1.2× slower than PyTorch’s addmm / matmul on MPS for 1280×1280 BF16 matrices. This GEMM shape is representative of Nano Chat–style transformer training, so the gap directly reduces end‑to‑end training throughput vs PyTorch.

It sped  up after NA support but its still behind since official Neural Accelerator Support

Repro
Gist with script and logs: https://gist.github.com/Anemll/5420800c3d29c7fae18a2b9b10907b14
Key details:
Shape: (1280, 1280)
Dtype: bfloat16
Device: PyTorch mps, MLX Metal
Warmup: 30, Iterations: 1000
Ops tested:
addmm: beta * C + alpha * (A @ B)
matmul: A @ B
Sync:
PyTorch: torch.mps.synchronize()
MLX: mx.eval(result)
Results (MLX 0.31.0)
PyTorch MPS:
addmm: ~0.596 ms
matmul: ~0.549 ms
MLX:
addmm: ~0.651 ms
matmul: ~0.646 ms
Ratios:
addmm: 1.09× slower (MLX / PyTorch)
matmul: 1.18× slower
Motivation (Nano Chat)
Small LLM (“Nano Chat”) training is GEMM‑bound at these sizes; this 10–20% gap in BF16 GEMM on M‑series Macs translates almost directly into slower training steps vs PyTorch MPS. Closing this gap would make MLX more competitive as the default backend for local Nano Chat training.

**Comment by jagrit06** [comment]
I will get to addressing this issue 
In the meanwhile, just so I also have an idea of the real world use case associated, do the tests you're doing with NanoChat also require the added `C` matrix to be full `1280x1280` ?

**Comment by Anemll** [comment]
Yes, the real-world use case requires full 1280×1280 C matrices (not bias vectors).

**NanoChat model uses `model_dim = 1280`** (20 layers × 64 aspect ratio). The Muon optimizer's Newton-Schulz iteration operates directly on weight matrices:

| Weight Matrix | Shape | NS iteration operates on |
|---|---|---|
| Attention Q/K/V/O | (1280, 1280) | 1280×1280 |
| MLP up-projection | (1280, 5120) | Transposed → 1280×1280 |
| MLP down-projection | (5120, 1280) | Transposed → 1280×1280 |

Each Newton-Schulz step (5 steps per optimizer update) does:
```python
A = X @ X.T                                        # (1280,1280) @ (1280,1280)
B = mx.addmm(b * A, A, A, beta=1.0, alpha=c)       # full matrix C, not a bias
X = mx.addmm(a * X, B, X, beta=1.0, alpha=1.0)     # full matrix C, not a bias
```

So the `C` argument to `addmm` is always a **full 1280×1280 matrix** (result of a prior computation), not a broadcast bias vector.

**Update on performance:** I just re-ran the benchmarks on MLX 0.30.0 and the gap has narrowed considerably from the ~3x I originally reported:

| Shape | PyTorch MPS (ms) | MLX (ms) | Ratio |
|---|---|---|---|
| 1280×1280 | 5.11 | 6.19 | 1.21x |
| 1280×320 | 0.82 | 1.02 | 1.24x |
| 5120×1280 | 15.19 | 17.02 | 1.12x |

**Average: ~1.19x**The `addmm` fusion itself is actually faster than PyTorch now — the remaining gap seems to come from plain matmul (`X @ X.T`).

**Comment by loading-awesome** [comment]
Another data point, from a different chip and a different shape regime — and on this one the gap looks like tile selection rather than kernel quality. MLX already has a Steel kernel that ties MPSGraph here; `GEMM_TPARAM_MACRO` just doesn't pick it at large M.

**Setup.** M3 Ultra (`applegpu_g15d`, 60-core), macOS 26.5.2, MLX 0.31.1 via mlx-swift 0.31.6. Compared against `MPSGraph.matrixMultiplication` directly rather than through PyTorch, bf16 in and bf16 out, same input bytes fed to both. 15 iterations per shape, the two arms interleaved. Outputs are **bit-identical** to MPSGraph at every shape below, so this is purely a scheduling difference.

**The observation.** At M = 4096 MLX matches MPSGraph. At M = 15,731 it falls 7–18% behind, and the deficit grows with K:

| M | K | N | MLX | MPSGraph | MPS/MLX |
|---:|---:|---:|---:|---:|---:|
| 4096 | 5376 | 4096 | 17.5 | 18.8 | 1.08x |
| 4096 | 14336 | 4096 | 18.9 | 18.9 | 1.00x |
| 4096 | 14336 | 5376 | 18.7 | 18.8 | 1.00x |
| 15731 | 2688 | 4096 | 18.1 | 19.4 | 1.07x |
| 15731 | 5376 | 5376 | 18.1 | 19.6 | 1.09x |
| 15731 | 14336 | 4096 | 16.7 | 19.6 | 1.17x |
| 15731 | 14336 | 5376 | 16.4 | 19.4 | 1.18x |

(TFLOP/s at the median. N is crossed at 4096 and 5376 to check it isn't an N effect — it isn't.)

**Cause.** These shapes land in the `devc == 'd'` branch of `GEMM_TPARAM_MACRO`, "large matmul / half and bfloat / reasonable K", which selects `64x64x16, wm=1 wn=2`. Sweeping the Steel configurations already compiled into the metallib, at the largest shape (15731x5376x28672):

| route | TFLOP/s |
|---|---:|
| `64x64x16, 1x2` (current) | 18.0–18.1 |
| `32x32x16, 2x2` | 18.3 |
| `64x32x32, 2x2` | 19.3 |
| `64x64x16, 2x2` | 19.6 |
| **`32x64x16, 1x2`** | **19.7–19.8** |

No new kernel — `32x64x16, 1x2` is already there, and is what the same branch selects for the "nn with large K" case a few lines below.

**Proposed change**, deliberately conservative — it only moves shapes at or above M = 8192, leaving the M = 4096 regime that already ties MPSGraph on the current route:

```diff
   } else if (devc == 'd') { /* Large device */
     if ((size_t)batch_size_out * M * N >= 1ul << 20) { /* large matmul */
       if (out.dtype() != float32) { /* half and bfloat */
         if (2 * std::max(M, N) > K) { /* Reasonable K */
-          bm = 64;
-          bn = 64;
-          bk = 16;
-          wm = 1;
-          wn = 2;
+          if (M >= 8192) { /* large-M tuning */
+            bm = 32;
+            bn = 64;
+            bk = 16;
+            wm = 1;
+            wn = 2;
+          } else {
+            bm = 64;
+            bn = 64;
+            bk = 16;
+            wm = 1;
+            wn = 2;
+          }
```

**After.** Every large-M cell ties MPSGraph, at both N and all three K:

| M | K | N | before | after | MPS |
|---:|---:|---:|---:|---:|---:|
| 15731 | 2688 | 4096 | 18.1 | **19.6** | 19.5 |
| 15731 | 5376 | 4096 | 18.0 | **19.7** | 19.6 |
| 15731 | 14336 | 4096 | 16.7 | **19.6** | 19.6 |
| 15731 | 2688 | 5376 | 18.2 | **19.6** | 19.6 |
| 15731 | 5376 | 5376 | 18.1 | **19.7** | 19.7 |
| 15731 | 14336 | 5376 | 16.4 | **19.3** | 19.5 |

M = 4096 rows are unchanged, as intended. fp16 takes the same branch and improves identically (19.6–19.8, tying MPSGraph). Output stays bit-identical throughout. In the workload this came from — a 50-block DiT at S = 15,731 — a full block goes from ~1290 ms to 1179 ms, about 8.6%.

**What I have not checked**, and would want to before this lands:

- `GEMM_TPARAM_MACRO` is also used by `gather_mm` and `segmented_mm`. Only `steel_matmul_regular_axpby` is exercised here.
- The M = 8192 threshold is measured at its endpoints (4096 below, 8192 at the gate); nothing in between.
- The branch is transpose-agnostic; only the combinations my workload uses were measured.
- One device. Whether `devc == 's'` wants the same treatment is untested, and the M5/NA path this issue originally reported is a different regime that I can't reproduce here.

Happy to open a PR with this, or to run additional shapes if that would be more useful than a patch — say which.

**Comment by loading-awesome** [comment]
Harness and patch, in case it's easier to reproduce than to re-derive: [loading-awesome/MiniMax-H3-Swift@dec6e92](https://github.com/loading-awesome/MiniMax-H3-Swift/commit/dec6e925c6d42052c57436c739c408e28fdb2014)

- [`Tests/H3ModulesTests/HardwareCeilingTests.swift`](https://github.com/loading-awesome/MiniMax-H3-Swift/blob/dec6e925c6d42052c57436c739c408e28fdb2014/Tests/H3ModulesTests/HardwareCeilingTests.swift) — `gemmKSweep` (the K sweep and the crossed M/N grid) and `mpsGraphBF16` (the bit-identity check, with a one-ULP control and a NaN-poisoned output buffer so a reported zero can't come from reading the wrong buffer)
- [`patches/mlx-m3-ultra-large-m-gemm.patch`](https://github.com/loading-awesome/MiniMax-H3-Swift/blob/dec6e925c6d42052c57436c739c408e28fdb2014/patches/mlx-m3-ultra-large-m-gemm.patch) — the routing change exactly as measured
- [`docs/PERF_ROADMAP.md`](https://github.com/loading-awesome/MiniMax-H3-Swift/blob/dec6e925c6d42052c57436c739c408e28fdb2014/docs/PERF_ROADMAP.md) §8, §13, §14 — the full before/after, the Steel configuration sweep, and the earlier dead ends

`H3_BIG=1 swift test --filter gemmKSweep` runs standalone — no model weights or checkpoint needed, just MLX and Metal.
